### ISSUE TITLE
Data set values on Horizontal Bar chart disappear on zooming along y-axis

### ISSUE THREAD

**Issue by charan1998** [PhilJay/MPAndroidChart#4131]
**Summary**

**Expected Behavior**

**Possible Solution**

**Device (please complete the following information):**
 - Device: [e.g. Google Pixel]
 - Android Version [e.g. 7.0]
 - Library Version (e.g. 3.0.3)

**Additional Context**

The values next to the bar in horizontal bar graph disappear on scaling along y-axis. It reappears when scrolled to far left

**Comment by Ezerous** [comment]
I am having the same issue...

**Comment by almic** [comment]
Could one of you please explain this issue further, perhaps with a video if you can?

**Comment by Ezerous** [comment]
Starting from [this](https://i.imgur.com/IVWdkrn.png) initially and zooming just a bit till zero/start goes off the chart makes the values on top of the bars disappear (as seen [here](https://i.imgur.com/TMojhzw.png)). While zoomed in, any time the start appears in the chart again, the values also reappear.

Deleting the useless parts, the minimal dummy code below can still reproduce the issue:

```java
HorizontalBarChart horizontalBarChart;                
List<BarEntry> barEntries = new ArrayList<>();
barEntries.add(new BarEntry(1, 1));
barEntries.add(new BarEntry(2, 2));
barEntries.add(new BarEntry(3, 3));
BarDataSet dataSet = new BarDataSet(barEntries, "Label");
YAxis yAxisLeft = horizontalBarChart.getAxisLeft();
yAxisLeft.setAxisMinimum(0);
BarData barData = new BarData(dataSet);
horizontalBarChart.setData(barData);
int chartHeightDp = 10 + 30 * entries.length;
DisplayMetrics metrics = context.getResources().getDisplayMetrics();
horizontalBarChart.setMinimumHeight((int) (chartHeightDp * (metrics.densityDpi / 160f)));
horizontalBarChart.invalidate();
```

Upon investigating it quickly, it seems like [this block of code](https://github.com/PhilJay/MPAndroidChart/blob/master/MPChartLib/src/main/java/com/github/mikephil/charting/renderer/HorizontalBarChartRenderer.java#L190) is triggered:
```java
if (!mViewPortHandler.isInBoundsX(buffer.buffer[j]))
  continue;
```
and the drawing of the values is skipped.

**Comment by almic** [comment]
Yeah that definitely looks like the cause. It should check the Y-value on horizontal bar charts instead of the X-value. I'll fix this soon.

**Comment by bernardo-macedo** [comment]
I'm using version v3.1.0 and seeing a very similar issue. But it is happening in the vertical version of the Bar chart. A few columns disappear in a given zoom level and drag position. Very weird.

Maybe it's related?

**Comment by parisisal** [comment]
I have this problem too.
I did a video.
[Android Emulator - Pixel_2_API_26_5554 2020-03-31 16-34-32.zip](https://github.com/PhilJay/MPAndroidChart/files/4409732/Android.Emulator.-.Pixel_2_API_26_5554.2020-03-31.16-34-32.zip)

**Comment by parisisal** [comment]
I found the problem.
The  BarEntry must added ordered (ascendent) by "X" value.
This code will generate the issue:

`ArrayList<BarEntry> binClassEntries = new ArrayList<BarEntry>();
for(int i = 0; i <20; i++)
    binClassEntries.add(new BarEntry((float)20-i,(float)new Random().nextInt(2000)));
`
This code will works:
`ArrayList<BarEntry> binClassEntries = new ArrayList<BarEntry>();
for(int i = 0; i <20; i++)
    binClassEntries.add(new BarEntry((float)i,(float)new Random().nextInt(2000)));`

**Comment by LukeDaniel16** [comment]
Any update or resolution?

**Comment by LukeDaniel16** [comment]
> I'm using version v3.1.0 and seeing a very similar issue. But it is happening in the vertical version of the Bar chart. A few columns disappear in a given zoom level and drag position. Very weird.
> 
> Maybe it's related?

You can resolve it?

**Comment by cocok30** [comment]
I got this problem, too...but it only happens when my inserted entries aren't in ascending order.

**Comment by RiotGoesWoof** [comment]
I have no idea why it works, but yeah. Sorting my values fixed the disappearing bars issue

**Comment by Razvann232** [comment]
I still didn't find any solution for the original issue. The best I could possibly do is disable scaling on the X axis (.setScaleXEnabled(false)).

**Comment by DaumasDev** [comment]
Based on the comments, I managed to formulate my solution to the rendering issue of bars when zooming. I created a class called 'DataXY' which has 2 attributes, 'valueX' and 'valueY'. Then, I iterate through the list of data that I want to display in the chart and add them to a 'mutableListOf<DataXY>()'. This option solves the case where there are different X values with different Y values, and the Y values need to be accumulated for the same X value. By creating the necessary logic, the data can be accumulated in a list of type 'DataXY'. If this list is sorted (ascending), the visual bug is resolved. After sorting this list, it is iterated through, and the data from the 'DataXY' objects are added to a new list of 'BarEntry', which is displayed in the chart.
